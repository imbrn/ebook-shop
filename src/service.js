let service;

export function setService(it) {
  service = it;
}

export function getService() {
  return service;
}
