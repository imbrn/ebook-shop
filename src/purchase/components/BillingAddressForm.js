import React, { PureComponent } from "react";
import { Formik } from "formik";
import StepPanel, { StepPanelField, TextField, Select } from "./StepPanel";
import nanoid from "nanoid";
import { getService } from "../../service";

export class BillingAddressForm extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidUpdate(prevProps) {
    const { billingAddress } = this.props;
    if (prevProps.billingAddress.zipCode !== billingAddress.zipCode) {
      this.autoFillAddress(billingAddress.zipCode);
    }
  }

  autoFillAddress(zipCode) {
    getService()
      .searchZipCode(zipCode)
      .then(result => {
        const { updateBillingAddress } = this.props;
        updateBillingAddress({
          zipCode: 123,
          state: result.state,
          city: result.city
        });
      })
      .catch(() => {
        updateBillingAddress({
          zipCode: "1234"
        });
        throw new Error("Error");
      });
  }

  render() {
    return <BillingAddressFormComponent {...this.props} />;
  }
}

const BillingAddressFormComponent = ({
  open,
  number,
  title,
  billingAddress,
  updateBillingAddress
}) => {
  const handleInputChange = (e, handleChange) => {
    updateBillingAddress({
      [e.target.name]: e.target.value
    });

    handleChange(e);
  };

  const zipCodeFieldId = nanoid(8);
  const stateFieldId = nanoid(8);
  const cityFieldId = nanoid(8);
  const addressFieldId = nanoid(8);

  return (
    <StepPanel open={open} number={number} title={title}>
      <Formik
        onSubmit={e => e.preventDefault()}
        initialValues={billingAddress}
        render={({
          values,
          touched,
          errors,
          handleChange,
          handleBlur,
          handleSubmit
        }) => (
          <form onSubmit={handleSubmit}>
            <StepPanelField
              component={TextField}
              fieldId={zipCodeFieldId}
              label="Zip code"
              placeholder="Your zip code here"
              name="zipCode"
              value={values.zipCode}
              error={touched.zipCode && errors.zipCode}
              onChange={e => handleInputChange(e, handleChange)}
              onBlur={handleBlur}
            />

            <StepPanelField
              component={Select}
              fieldId={stateFieldId}
              label="State"
              placeholder="Your state here"
              name="state"
              options={states}
              value={values.state}
              error={touched.state && errors.state}
              onChange={e => handleInputChange(e, handleChange)}
              onBlur={handleBlur}
            />

            <StepPanelField
              component={TextField}
              fieldId={cityFieldId}
              label="City"
              placeholder="Your city here"
              name="city"
              value={values.city}
              error={touched.city && errors.city}
              onChange={e => handleInputChange(e, handleChange)}
              onBlur={handleBlur}
            />

            <StepPanelField
              component={TextField}
              fieldId={addressFieldId}
              label="Address"
              placeholder="Your full address here"
              name="address"
              value={values.address}
              error={touched.address && errors.address}
              onChange={e => handleInputChange(e, handleChange)}
              onBlur={handleBlur}
            />
          </form>
        )}
      />
    </StepPanel>
  );
};

BillingAddressFormComponent.defaultProps = {
  open: true,
  number: 2,
  title: "Billing Address",
  billingAddress: {
    zipCode: "",
    state: "",
    city: "",
    address: ""
  },
  updateBillingAddress: () => {}
};

const states = [
  { label: "Acre", value: "AC" },
  { label: "Alagoas", value: "AL" },
  { label: "Amapá", value: "AP" },
  { label: "Amazonas", value: "AM" },
  { label: "Bahia", value: "BA" },
  { label: "Ceará", value: "CE" },
  { label: "Distrito Federal", value: "DF" },
  { label: "Espírito Santo", value: "ES" },
  { label: "Goiás", value: "GO" },
  { label: "Maranhão", value: "MA" },
  { label: "Mato Grosso", value: "MT" },
  { label: "Mato Grosso do Sul", value: "MS" },
  { label: "Minas Gerais", value: "MG" },
  { label: "Pará", value: "PA" },
  { label: "Paraíba", value: "PB" },
  { label: "Paraná", value: "PR" },
  { label: "Pernambuco", value: "PE" },
  { label: "Piauí", value: "PI" },
  { label: "Rio de Janeiro", value: "RJ" },
  { label: "Rio Grande do Norte", value: "RN" },
  { label: "Rio Grande do Sul", value: "RS" },
  { label: "Rondônia", value: "RO" },
  { label: "Roraima", value: "RR" },
  { label: "Santa Catarina", value: "SC" },
  { label: "São Paulo", value: "SP" },
  { label: "Sergipe", value: "SE" },
  { label: "Tocantins", value: "TO" }
];

export default BillingAddressForm;
