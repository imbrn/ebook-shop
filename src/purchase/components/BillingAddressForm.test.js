import React from "react";
import { render, cleanup, fireEvent } from "react-testing-library";
import { BillingAddressForm } from "./BillingAddressForm";
import { setService } from "../../service";

jest.mock("nanoid", () => {
  let nextId = 1;
  return () => nextId++;
});

jest.mock("react-select", () => {
  return ({ ...rest }) => <input type="text" {...rest} />;
});

afterEach(cleanup);

beforeEach(() => {
  setService({});
});

test("renders correctly", () => {
  const billingAddress = {
    zipCode: "14530000",
    state: "S찾o Paulo",
    city: "Miguel처polis",
    address: "Av. Leopoldo Carlos de Oliveira, 1234"
  };
  const { container } = render(
    <BillingAddressForm billingAddress={billingAddress} />
  );
  expect(container.firstChild).toMatchSnapshot();
});

test("updateBillingAddress", () => {
  const updateBillingAddress = jest.fn();
  const { getByPlaceholderText } = render(
    <BillingAddressForm updateBillingAddress={updateBillingAddress} />
  );

  fireEvent.change(getByPlaceholderText("Your zip code here"), {
    target: { value: "14530000" }
  });

  fireEvent.change(getByPlaceholderText("Your state here"), {
    target: { value: "SP" }
  });

  fireEvent.change(getByPlaceholderText("Your city here"), {
    target: { value: "Miguel처polis" }
  });

  fireEvent.change(getByPlaceholderText("Your full address here"), {
    target: { value: "My Street, 123" }
  });

  expect(updateBillingAddress.mock.calls).toEqual([
    [{ zipCode: "14530000" }],
    [{ state: "SP" }],
    [{ city: "Miguel처polis" }],
    [{ address: "My Street, 123" }]
  ]);
});
