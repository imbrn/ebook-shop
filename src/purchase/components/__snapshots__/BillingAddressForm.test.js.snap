// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <header>
    <span
      class=""
    >
      2
    </span>
    <span
      class=""
    >
      Billing Address
    </span>
  </header>
  <div>
    <form>
      <div>
        <div>
          <label
            class=""
            for="1"
          >
            Zip code
          </label>
        </div>
        <div>
          <input
            class=""
            id="1"
            name="zipCode"
            placeholder="Your zip code here"
            type="text"
            value="14530000"
          />
        </div>
      </div>
      <div>
        <div>
          <label
            class=""
            for="2"
          >
            State
          </label>
        </div>
        <div>
          <input
            id="2"
            name="state"
            options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
            placeholder="Your state here"
            type="text"
            value="São Paulo"
          />
        </div>
      </div>
      <div>
        <div>
          <label
            class=""
            for="3"
          >
            City
          </label>
        </div>
        <div>
          <input
            class=""
            id="3"
            name="city"
            placeholder="Your city here"
            type="text"
            value="Miguelópolis"
          />
        </div>
      </div>
      <div>
        <div>
          <label
            class=""
            for="4"
          >
            Address
          </label>
        </div>
        <div>
          <input
            class=""
            id="4"
            name="address"
            placeholder="Your full address here"
            type="text"
            value="Av. Leopoldo Carlos de Oliveira, 1234"
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;
